[4단계 - 동시성 확장하기] 플린트(이주환) 미션 제출합니다.

jhan0121 opened the pull request:
안녕하세요 돔푸! 이제 마지막 step이네요!

이번 step은 `ThreadPoolManager`와 `Connector` 부분을 위주로 확인해주시면 될 것 같습니다~

구현 자체는 어려운 점은 없었던 것 같고, 새로운 개념이 많이 나와서 이에 대한 학습 위주로 진행했습니다!

## Thread pool 관련 주요 설정 값
### core pool size
스레드 풀이 평상시에 유지하는 최소 스레드 개수

was는 i/o bound 작업 위주이므로, cpu 코어의 약 2배 크기로 설정

### maximum pool size
어떤 상황에서도 최대로 가질 수 있는 스레드의 총 개수

core pool size의 2배 크기로 설정

### wait queue size
core thread count 만큼의 스레드가 모두 일하고 있을 때 저장할 수 있는 큐의 최대 크기

큐의 크기를 명시하지 않으면 `Integer.MAX_VALUE` 크기로 자동 생성됨. 따라서 과도한 요청에 대한 바운더리를 지정해주어야 함.

Commit:
fix(study): Spring Boot 의존성 관리로 JUnit 정렬 및 테스트 오류 해결

- junit-jupiter-api/engine 직접 버전 제거
- spring-boot-starter-test로 테스트 의존성 일원화
- AssertJ/Mockito 버전 지정 제거해 Boot 관리로 위임
- JUnit Jupiter/Platform 불일치로 인한 NoSuchMethodError 해소

jhan0121 committed:
Merge branch 'woowacourse:jhan0121' into jhan0121

jhan0121 committed:
Merge branch 'woowacourse:jhan0121' into jhan0121

jhan0121 committed:
feat: Connector thread pool 적용

jhan0121 committed:
refactor: 동시성 처리 안전 컬렉션으로 변경

- ControllerProvider
- Session

jhan0121 committed:
Merge branch 'woowacourse:jhan0121' into jhan0121

jhan0121 committed:
refactor: 스레드 풀 생성 로직 ThreadPoolManager 로 분리

jhan0121 committed:
fix: ThreadPoolManager#createDefaultThreadPoolExecutor time unit 수정

jhan0121 committed:
fix: ThreadPoolManager time unit 설정 불일치 수정

jhan0121 committed:
fix: ThreadPoolManager 코어 스레드 수 & 최대 스레드 수 검증 로직 수정

dompoo commented on tomcat/src/main/java/org/apache/catalina/connector/ThreadPoolManager.java:
```diff
@@ -0,0 +1,71 @@
+
+    private static final Logger log = LoggerFactory.getLogger(ThreadPoolManager.class);
+
+    private static final int DEFAULT_CORE_THREAD_COUNT = 32;
+    private static final int DEFAULT_MAX_THREAD_COUNT = 64;
+    private static final long DEFAULT_KEEP_ALIVE_SECONDS = 60;
+    private static final int DEFAULT_MAX_WAIT_QUEUE_SIZE = 100;
```
오 이런 설정값은 정확히 어떻게 산출된건가요?
설명에 있긴 한데, 좀 더 자세하게 설명해주세요!!

jhan0121 replied:
### core thread count
core thread count의 경우에는 context switch를 통해 "잠깐씩 연산하고 I/O 대기하고..." 흐름이 각 스레드마다 반복하므로 스레드가 대기하는 동안 다른 스레드가 CPU를 사용하는 것이 더 유리하다는 생각이 들었습니다. 여기서 cpu 코어 수보다 더 많은 스레드를 만드는 것이 더 유리하다고 판단했습니다.
 cpu 코어의 2배로 잡은 것은 큰 이유는 없고, "코어마다 스레드 2개 정도씩 계속 가지고 있어도 충분히 처리할 수 있는 정도"라고 생각했습니다.

---

### max thread count
max thread count의 경우에는 "갑자기 많은 요청이 들어왔을 때, 완충 역할"을 한다고 생각했습니다. core의 2배 정도의 차이라면 "스레드 생성 오버헤드 부담"과 "많은 요청 처리 효율" 사이의 밸런스를 잡은 것 같다는 생각이 들어 적용해보았습니다.

---

### keep alive seconds
keep alive seconds의 경우에는 "core count보다 더 많이 생긴 스레드를 정리하기 시작하는 시간"이라고 보았습니다. 1분 정도라면 요청이 줄었다가 다시 늘어나는 경우를 어느 정도 커버할 수 있다고 보았습니다.

---

### max wait queue size
max wait queue size의 경우에는 `Connector`의 `DEFAULT_ACCEPT_COUNT`와 같게 하면 os와 was 어느 한쪽으로 부하가 쏠리지 않을 것 같아서 안정성과 제어 용이성을 적절히 챙길 수 있을 것 같아 적용해보았습니다.

dompoo replied:
좋네유~

dompoo commented on tomcat/src/main/java/org/apache/catalina/connector/ThreadPoolManager.java:
```diff
@@ -0,0 +1,71 @@
+
+    private static int checkMaxThreadCount(final int maxThreads, final int coreThreads) {
+        if (coreThreads > maxThreads) {
+            log.warn(
+                    "detected core counts bigger than max count: change max count to bigger value between maxCount={} and coreCount={}",
+                    maxThreads, coreThreads);
```
뭐에요 이거 AI 썼죠

jhan0121 replied:
ㅋㅋㅋㅋㅋ메세지는 직접 작성했어요! 코드 래빗 코멘트에 core count 와 max count 검증을 더 추가해야 하는 것을 추천해서 적용해보았습니다!

dompoo commented on tomcat/src/main/java/org/apache/catalina/session/Session.java:
```diff
@@ -4,13 +4,13 @@
 
     private final String id;
-    private final Map<String, Object> values = new HashMap<>();
+    private final Map<String, Object> values = new ConcurrentHashMap<>();
```
Session에서 동시성을 고려하신 것 같은데, 그 이유가 뭔가요?

jhan0121 replied:
흔한 일은 아닐 것 같긴 하지만 "다른 곳에서 같은 session에 연결된 http 요청이 동시에 session을 만지는 일은 없을까?"하는 생각에서 시작했어요!

그러다보면 누락되는 session 값이 있을 것 같아 적용해보았습니다!

dompoo replied:
방어적으로 개발했다고 봐도 되겠네요! 좋습니다.

dompoo commented on tomcat/src/main/java/org/apache/coyote/http11/controller/ControllerProvider.java:
```diff
@@ -10,7 +10,7 @@ public enum ControllerProvider {
 
     ControllerProvider() {
-        this.controllers = new ArrayList<>();
+        this.controllers = new CopyOnWriteArrayList<>();
```
컨트롤러도 마찬가지 입니다! 사용하신 이유가 궁금해요.

jhan0121 replied:
컨트롤러의 경우에는 구성 로직이 잘못된 형태였네요...!

기존
```
tomcat 초기화  -> 연결이 들어오면(tomcat.process() 호출 시) -> Http11Processor에서 DefaultControllerConfig가 기본 controller를 controllerProvider에 삽입
```

여기서 `process()`를 할 때마다 컨트롤러를 초기화 하는 것부터가 문제였습니다. 이러면 사용자 연결이 생길 때마다 기본 컨트롤러가 계속 삽입되는 문제가 있네요.

따라서 Tomcat이 생성될 때에, 최초 생성하도록 수정하였고, 이 구조에서는 메인 스레드 하나에서만 실행되기 때문에 동시성 컬렉션은 의미가 없어져 ArrayList로 다시 롤백했습니다.

https://github.com/woowacourse/java-http/pull/1034/commits/54f08019f95aa13021031fa981113f5a5ec1a6e8

dompoo reviewed (changes requested):
안녕하세요 플린트~ 이번 스텝은 간단한 편이라서 질문만 몇개 남겼습니다!

jhan0121 committed:
refactor: DefaultControllerConfig 초기화 시점 이동

- Http11Processor -> Tomcat 으로 이동하여 최초 실행에만 초기화 하도록 로직 변경
- ControllerProvider 일반 컬렉션으로 변경

jhan0121 commented:
아직 push되지 않았던 commit이 있었네요...! 바로 push 진행하고 대상 코멘트에 커밋 로그 남겼습니다. 불편을 드려 죄송합니다ㅠㅠ

dompoo reviewed (approved):
안뇽하세요 플린트~ 마지막 리뷰가 늦었네요!
전체적으로 코드 잘 작성해주셔서 이번 미션은 여기까지 하면 될 것 같아요.
고생하셨고 다음에 오프에서 보죠 ㅎㅎ
